Detectors/CPV/calib/CPVCalibWorkflow/src/CPVGainCalibDevice.cxx

```diff
@@ -0,0 +1,231 @@
+// Copyright CERN and copyright holders of ALICE O2. This software is
+// distributed under the terms of the GNU General Public License v3 (GPL
+// Version 3), copied verbatim in the file "COPYING".
+//
+// See http://alice-o2.web.cern.ch/license for full licensing information.
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+#include "CPVCalibWorkflow/CPVGainCalibDevice.h"
+#include "CCDB/CcdbApi.h"
+#include "CCDB/CcdbObjectInfo.h"
+#include <string>
+#include <ctime>
+#include "FairLogger.h"
+#include "CommonDataFormat/InteractionRecord.h"
+#include "Framework/ConfigParamRegistry.h"
+#include "Framework/ControlService.h"
+#include "Framework/WorkflowSpec.h"
+#include "Framework/InputRecordWalker.h"
+#include "DetectorsRaw/RDHUtils.h"
+#include "DetectorsCalibration/Utils.h"
+#include "CCDB/CcdbObjectInfo.h"
+#include "CommonUtils/MemFileHelper.h"
+#include "DataFormatsCPV/TriggerRecord.h"
+#include "CPVReconstruction/RawReaderMemory.h"
+#include "CPVReconstruction/RawDecoder.h"
+#include "CPVBase/Geometry.h"
+#include "TF1.h"
+
+using namespace o2::cpv;
+
+void CPVGainCalibDevice::init(o2::framework::InitContext& ic)
+{
+  //Check if files from previous runs exist
+  //if yes, read histogram
+  mMean = std::unique_ptr<TH2F>(new TH2F("Gains", "Signals per channel", o2::cpv::Geometry::kNCHANNELS, 0.5, o2::cpv::Geometry::kNCHANNELS + 0.5, 1024, 0., 4096.));
+
+  std::string filename = mPath + "CPVGains.root";
+  TFile filein(filename.data(), "READ");
+  if (filein.IsOpen()) {
+    mMean->Add(static_cast<TH2F*>(filein.Get("Gains")));
+    filein.Close();
+  }
```

Comment:
You are reading the file into the histo, filling the latter and save it again, is this in order to accumulate over multiple runs? In this way you don't have a control over what you get from the disk, if such an accumulation is needed, it should be done either via CCDB, or you via filling containers which are then merged and processed.

Reply:
Yes, this calibration may require scan of several runs. These are transient data with relatively large size, probably CCDB is not good place to store them. Files with histos can be used as "containers" and merged if necessary with root tools. I plan to provide some configuration with paths to files etc. Or we have this functionality implemented in framework?

Reply:
Concerning CCDB, I meant to put there not transient histos but to incrementally cumulate the calibration, if the algorithm allows this. How are you going to control the period over which the integration goes? 
Also, do you really need to update the CCDB in the end of the workflow, if in the next session you still keep updating this file? Are these objects used for the sync. workflow?
Eventually, the containers will be handled by the framework. Anyway, the changes can be itnroduced later.

Reply:
For gain-like calibrations there is no much sense in updating with intermediate statistics, so only final objects should be sent. I plan to check mean statistics per channel and finalize calibration once threshold is set. This will probably require manual intervention, but this should be rare procedure ~ 1 time per year. Normally CCDB will not be updated per processing at EOS, only temporary objects will be stored to files. Only when statistics will be reached calibration will be send to CCDB. Output to QC will be send on each end of workflow and experts will be able to look at status. 
For PHOS some versions of gain will be used in reconstruction but final calibration will be applied on analysis level. For CPV gain calibration will be necessary for clustering. But it should be produced on commissioning stage.